Issue 6904 first free council vote

shamb0 opened the pull request:
**Beta implementation to resolve the issue [Participating in Governance should be Free for First Time Voters #6904](https://github.com/paritytech/substrate/issues/6904)**

---

## First level of demo sequence

![](https://i.imgur.com/zt4iSyS.jpg)

shamb0 commented:
Hello Shawn,


Like to do quick check on implementation approach for the issue.

In above Demo test sequence, able the detect the first-time voters. But looking for some clarity on balance transaction.
From my initial analysis, looks like fee is not getting detected in all cases ...

shamb0 commented:
Hello Shawn,

Thanks for guidance, reverted pallet-democracy changes.
and modified only pallet-collective.

Used below logic to detect the first vote within a proposal.

![](https://i.imgur.com/u9OW1uK.png)

And verified with testcase "tests::motions_revoting_works()" [slightly modified] ...

From below trace account 2 is detected as first vote in a proposal ...

![](https://i.imgur.com/bAOEwQU.png)

**Note:: Plz ignore changes in "frame/democracy/src/tests.rs", Some mistake in my cleanup ... will correct it :-)**

shawntabrizi commented on frame/collective/src/lib.rs:
```diff
@@ -505,6 +518,7 @@ decl_module! {
 			#[compact] index: ProposalIndex,
 			approve: bool,
 		) -> DispatchResultWithPostInfo {
+			let mut is_account_voting_first_time: bool = false;
```
```suggestion
```

shawntabrizi replied:
Dont use mutable variables like this if you dont need to

shamb0 replied:
Thanks Shawn will correct my bad coding style ...

shawntabrizi commented on frame/collective/src/lib.rs:
```diff
@@ -515,6 +529,22 @@ decl_module! {
 			let position_no = voting.nays.iter().position(|a| a == &who);
 
+			// first vote of member in the motion
+			if position_yes.is_none() && position_no.is_none() {
```
```suggestion
			let is_account_voting_first_time = position_yes.is_none() && position_no.is_none();
```

shawntabrizi commented on frame/collective/src/lib.rs:
```diff
@@ -515,6 +529,22 @@ decl_module! {
 			let position_yes = voting.ayes.iter().position(|a| a == &who);
 			let position_no = voting.nays.iter().position(|a| a == &who);
 
+			// first vote of member in the motion
+			if position_yes.is_none() && position_no.is_none() {
+				is_account_voting_first_time = true;
+			}
```
```suggestion
			// first vote of member in the motion
			let is_account_voting_first_time = position_yes.is_none() && position_no.is_none();
```

shawntabrizi commented on frame/collective/src/lib.rs:
```diff
@@ -537,11 +567,29 @@ decl_module! {
 
-			Ok(Some(T::WeightInfo::vote(members.len() as u32)).into())
+			if is_account_voting_first_time {
+				Ok(Pays::No.into())
```
You need weight info

shamb0 replied:
Fine, Corrected it ...

shawntabrizi commented on frame/collective/src/lib.rs:
```diff
@@ -537,11 +564,33 @@ decl_module! {
 
-			Ok(Some(T::WeightInfo::vote(members.len() as u32)).into())
+			if is_account_voting_first_time {
+				Ok((Some(
+					T::WeightInfo::vote(members.len() as u32)),
+					Pays::No).into())
+			}
+			else {
+				Ok((Some(
+					T::WeightInfo::vote(members.len() as u32)),
+					Pays::Yes).into())
+			}
```
```suggestion
				Ok((
					Some(T::WeightInfo::vote(members.len() as u32)),
					Pays::No,
				).into())
			}
			else {
				Ok((
					Some(T::WeightInfo::vote(members.len() as u32)),
					Pays::Yes,
				).into())
			}
```

shamb0 commented:
Hello Shawn,

Moved the "mod test" to tests.rs

Introduced new test case "motions_all_first_vote_free_works()" to validate the changes ...

looks like I am messing up in testing for Pay::Yes & Pay::No

For Ok condition I using assert_eq!() pattern

![](https://i.imgur.com/V23CxoJ.png)

And for Err case I am using match {} pattern

![](https://i.imgur.com/j4AsE09.png)

Plz suggest best alternate ...

shamb0 commented:
Plz Ignore this PR, instead use  [ PR | 7661 ](https://github.com/paritytech/substrate/pull/7661)
